server/main.ts

```diff
@@ -0,0 +1,156 @@
+import http from "http";
+import Koa from "koa";
+import Router from "koa-router";
+import { JSDOM } from "jsdom";
+import XYChart from "../packages/xy-chart";
+import {
+  convertStarDataToChartData,
+  DEFAULT_MAX_REQUEST_AMOUNT,
+  getReposStarData,
+} from "../common/chart";
+import api from "../common/api";
+import cache from "./cache";
+import { replaceSVGContentFilterWithCamelcase } from "./utils";
+import { getNextToken, initTokenFromEnv } from "./token";
+import { ChartMode } from "../types/chart";
+import { CHART_TYPES } from "./const";
+
+const startServer = async () => {
+  await initTokenFromEnv();
+
+  const app = new Koa();
+  const router = new Router();
+
+  // Example request link:
+  // /svg?repos=bytebase/star-history&type=Date
+  router.get("/svg", async (ctx) => {
+    const repos = `${ctx.query["repos"]}`.split(",");
+    let type = `${ctx.query["type"]}` as ChartMode;
+
+    if (!CHART_TYPES.includes(type)) {
+      type = "Date";
+    }
+
+    if (repos.length === 0) {
+      ctx.throw(400, `${http.STATUS_CODES[400]}: Repo name required`);
+      return;
+    }
+
+    const token = getNextToken();
+    const reposStarData = [];
+    const nodataRepos = [];
+
+    for (const repo of repos) {
+      if (cache.has(repo)) {
```

Comment:
You may check its API if calling cache.has great advantage. Otherwise below is more efficient (1 less lookup)

const cacheData = cache.get(repo)

if (cacheData) {
} else {
}